app/ts/library/utilities.ts

```diff
@@ -20,3 +22,30 @@ export function removeNonStringsAndTrim(...strings: (string | boolean | undefine
 			.join(' ')
 	)
 }
+
+/**
+ * Check string for valid address format
+ */
+export function assertsAddress(string: string): asserts string is `0x${string}` {
+	if (!/^0x[0-9A-Fa-f]{40}$/.test(string)) throw new Error(`Invalid address ${string}`)
+}
+
+export function isAddress(string: string): string is `0x${string}` {
+	return /^0x[0-9A-Fa-f]{40}$/.test(string)
+}
+
+/**
+ * Describe a window ethereum object
+ */
+type ExternalProvider = ethers.providers.ExternalProvider
+export function assertsExternalProvider(ethereum: unknown): asserts ethereum is ExternalProvider {
+	if (ethereum === null || ethereum === undefined || typeof ethereum !== 'object') throw new Error('Ethereum object does not exist')
+}
+
+export type EthereumWithListeners = {
```

Comment:
no longer needed after we use ethers directly

Reply:
this is still needed to narrow down ethers `provider.provider` to ensure the `on` method is present. `ExternalProvider` does not include one by default